src/pykeen/sampling/pseudo_type.py

```diff
@@ -0,0 +1,75 @@
+"""Pseudo-Typed negative sampling."""
+import logging
+import random
+from collections import defaultdict
+from typing import Optional, Tuple
+
+import torch
+
+from .negative_sampler import NegativeSampler
+from ..triples import CoreTriplesFactory
+
+__all__ = [
+    "PseudoTypedNegativeSampler",
+]
+
+logger = logging.getLogger(__name__)
+
+
+class PseudoTypedNegativeSampler(NegativeSampler):
+    """
+    A negative sampler using pseudo-types.
+
+    To generate a corrupted head entity for triple (h, r, t), only those entities are considered which occur as a
+    head entity in a triple with the relation r.
+    """
+
+    def __init__(
+        self,
+        *,
+        triples_factory: CoreTriplesFactory,
+        **kwargs,
+    ):
+        """
+        Instantiate the negative sampler.
+
+        :param triples_factory:
+            The factory holding the positive training triples
+        :param kwargs:
+            Additional keyword based arguments passed to NegativeSampler.
+        """
+        super().__init__(triples_factory=triples_factory, **kwargs)
+        # index triples
+        self.tails = defaultdict(set)
+        self.heads = defaultdict(set)
+        for h, r, t in triples_factory.mapped_triples.tolist():
```

Comment:
externalize this indexing in a function in `pykeen.triples.analysis` with the other similar ones we did in #390